File: host/src/main/java/dk/dtu/domain/rules/effects/TileEffect.java
package dk.dtu.domain.rules.effects;


import dk.dtu.domain.core.Phase;
import dk.dtu.domain.model.Tile;
import dk.dtu.domain.rules.api.BoardAPI;

import java.util.EnumSet;

/**
 * A tile effect that can activate during specific game phases.
 * <p>
 * Each {@code TileEffect} defines one or more {@link Phase phases} during which
 * it triggers and an {@link #onPhase(Phase, Tile, BoardAPI)} method that applies
 * its behavior to the tile and any robots or elements on it.
 * </p>
 * @author Weihao Mo
 * @author William Pii Jæger
 */
public interface TileEffect {
    void onPhase(Phase phase, Tile tile, BoardAPI api);
    EnumSet<Phase> phases();

    default boolean triggersOn(Phase p) {
        return phases().contains(p);
    }
}

File: engine/src/effects/tileEffect.ts
import type { Phase } from "../core/phase.js";
import type { Tile } from "../model/tile.js";
import type { BoardAPI } from "../rules/boardApi.js";

/** Ported from dk.dtu.domain.rules.effects.TileEffect. */
export interface TileEffect {
  onPhase(phase: Phase, tile: Tile, api: BoardAPI): void;
  /** Phases this effect triggers on, or null when it has no phase behaviour. */
  phases(): Set<Phase> | null;
}

export function triggersOn(effect: TileEffect, p: Phase): boolean {
  const phases = effect.phases();
  return phases !== null && phases.has(p);
}
